server/sonar-server/src/main/java/org/sonar/server/component/ws/SearchProjectsAction.java

```diff
@@ -0,0 +1,132 @@
+import org.sonarqube.ws.Common;
+import org.sonarqube.ws.WsComponents.Component;
+import org.sonarqube.ws.WsComponents.SearchProjectsWsResponse;
+import org.sonarqube.ws.client.component.SearchProjectsRequest;
+
+import static org.sonar.api.utils.Paging.offset;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.component.SearchProjectsRequest.DEFAULT_PAGE_SIZE;
+import static org.sonarqube.ws.client.component.SearchProjectsRequest.MAX_PAGE_SIZE;
+
+public class SearchProjectsAction implements ComponentsWsAction {
+  private final DbClient dbClient;
+
+  public SearchProjectsAction(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    context.createAction("search_projects")
+      .setSince("6.2")
+      .setDescription("Search for projects")
+      .addPagingParams(DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE)
+      .setInternal(true)
+      .setResponseExample(getClass().getResource("search_projects-example.json"))
+      .setHandler(this);
+  }
+
+  @Override
+  public void handle(Request httpRequest, Response httpResponse) throws Exception {
+    SearchProjectsWsResponse response = doHandle(toRequest(httpRequest));
+
+    writeProtobuf(response, httpRequest, httpResponse);
+  }
+
+  private SearchProjectsWsResponse doHandle(SearchProjectsRequest request) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      SearchResult searchResult = searchProjects(dbSession, request);
+
+      return buildResponse(request, searchResult);
+
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private SearchResult searchProjects(DbSession dbSession, SearchProjectsRequest request) {
+    List<ComponentDto> allProjects = dbClient.componentDao().selectProjects(dbSession);
+    List<ComponentDto> projects = allProjects
+      .stream()
+      .sorted((c1, c2) -> c1.name().compareTo(c2.name()))
+      .skip(offset(request.getPage(), request.getPageSize()))
+      .limit(request.getPageSize())
+      .collect(Collectors.toList());
+
+    return new SearchResult(projects, allProjects.size());
+  }
+
+  private static SearchProjectsRequest toRequest(Request httpRequest) {
+    SearchProjectsRequest.Builder request = SearchProjectsRequest.builder();
+
+    request.setPage(httpRequest.mandatoryParamAsInt(Param.PAGE));
+    request.setPageSize(httpRequest.mandatoryParamAsInt(Param.PAGE_SIZE));
+
+    return request.build();
+  }
+
+  private static SearchProjectsWsResponse buildResponse(SearchProjectsRequest request, SearchResult searchResult) {
+    SearchProjectsWsResponse.Builder response = SearchProjectsWsResponse.newBuilder();
+
+    response.setPaging(Common.Paging.newBuilder()
+      .setPageIndex(request.getPage())
+      .setPageSize(request.getPageSize())
+      .setTotal(searchResult.total));
+
+    Function<ComponentDto, Component.Builder> dbToWsComponent = dbComponent -> Component.newBuilder()
```

Comment:
rather than a lambda assigned to a variable, use a method and a method reference
it will make `buildResponse` shorter, avoids creating a lamba object and assigning it to a variable.
methods can be documented with javadoc if need be